render/render.go

```diff
@@ -1,203 +1,121 @@
-	"encoding/json"
-	"encoding/xml"
-	"fmt"
 	"net/http"
 	"reflect"
 )
 
-// Respond handles streaming JSON and XML responses, automatically setting the
-// Content-Type based on request headers. It will default to a JSON response.
-func Respond(w http.ResponseWriter, r *http.Request, v interface{}) {
-	// Present the object.
-	if presenter, ok := r.Context().Value(presenterCtxKey).(Presenter); ok {
-		r, v = presenter.Present(r, v)
-	}
-
-	switch reflect.TypeOf(v).Kind() {
-	case reflect.Chan:
-		switch getResponseContentType(r) {
-		case ContentTypeEventStream:
-			channelEventStream(w, r, v)
-			return
-		default:
-			v = channelIntoSlice(w, r, v)
-		}
-	}
+type Renderer interface {
+	Render(w http.ResponseWriter, r *http.Request) error
+}
 
-	// Format data based on Content-Type.
-	switch getResponseContentType(r) {
-	case ContentTypeJSON:
-		JSON(w, r, v)
-	case ContentTypeXML:
-		XML(w, r, v)
-	default:
-		JSON(w, r, v)
-	}
+type Binder interface {
+	Bind(r *http.Request) error
 }
 
-// PlainText writes a string to the response, setting the Content-Type as
-// text/plain.
-func PlainText(w http.ResponseWriter, r *http.Request, v string) {
-	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
-	if status, ok := r.Context().Value(statusCtxKey).(int); ok {
-		w.WriteHeader(status)
+func Bind(r *http.Request, v Binder) error {
+	if err := Decode(r, v); err != nil {
+		return err
 	}
-	w.Write([]byte(v))
+	return binder(r, v)
 }
 
-// Data writes raw bytes to the response, setting the Content-Type as
-// application/octet-stream.
-func Data(w http.ResponseWriter, r *http.Request, v []byte) {
-	w.Header().Set("Content-Type", "application/octet-stream")
-	if status, ok := r.Context().Value(statusCtxKey).(int); ok {
-		w.WriteHeader(status)
+func Render(w http.ResponseWriter, r *http.Request, v Renderer) error {
+	if err := renderer(w, r, v); err != nil {
+		return err
 	}
-	w.Write(v)
+	Respond(w, r, v)
+	return nil
 }
 
-// HTML writes a string to the response, setting the Content-Type as text/html.
-func HTML(w http.ResponseWriter, r *http.Request, v string) {
-	w.Header().Set("Content-Type", "text/html; charset=utf-8")
-	if status, ok := r.Context().Value(statusCtxKey).(int); ok {
-		w.WriteHeader(status)
+func RenderList(w http.ResponseWriter, r *http.Request, l []Renderer) error {
+	for _, v := range l {
+		if err := renderer(w, r, v); err != nil {
+			return err
+		}
 	}
-	w.Write([]byte(v))
+	Respond(w, r, l)
+	return nil
 }
 
-// JSON marshals 'v' to JSON, automatically escaping HTML and setting the
-// Content-Type as application/json.
-func JSON(w http.ResponseWriter, r *http.Request, v interface{}) {
-	w.Header().Set("Content-Type", "application/json; charset=utf-8")
-	if status, ok := r.Context().Value(statusCtxKey).(int); ok {
-		w.WriteHeader(status)
+// Executed top-down
+func renderer(w http.ResponseWriter, r *http.Request, v Renderer) error {
+	rv := reflect.ValueOf(v)
+	if rv.Kind() == reflect.Ptr {
+		rv = rv.Elem()
```

Comment:
My reflect knowledge is somewhat lacking: why not simply `v = reflect.Indirect(reflect.ValueOf(v))` ?